clustpy/deep/dkm.py

```diff
@@ -0,0 +1,543 @@
+    # Use DKM optimizer parameters (usually learning rate is reduced by a magnitude of 10)
+    optimizer = optimizer_class(list(autoencoder.parameters()) + list(dkm_module.parameters()),
+                                **clustering_optimizer_params)
+    # DKM Training loop
+    dkm_module.fit(autoencoder, trainloader, clustering_epochs, device, optimizer, loss_fn, cluster_loss_weight)
+    # Get labels
+    dkm_labels = predict_batchwise(testloader, autoencoder, dkm_module, device)
+    dkm_centers = dkm_module.centers.detach().cpu().numpy()
+    # Do reclustering with Kmeans
+    embedded_data = encode_batchwise(testloader, autoencoder, device)
+    kmeans = KMeans(n_clusters=n_clusters, random_state=random_state)
+    kmeans.fit(embedded_data)
+    return kmeans.labels_, kmeans.cluster_centers_, dkm_labels, dkm_centers, autoencoder
+
+
+def _get_default_alphas(init_alpha: float = 0.1, n_alphas: int = 40) -> list:
+    """
+    Return a list with alphas equal to \alpha_{i+1}=2^{1/log(i)^2}*\alpha_i, where \alpha_1=0.1 and maximum i=40
+
+    Parameters
+    ----------
+    init_alpha : float
+        the initial alpha value (default: 0.1)
+    n_alphas : int
+        number of alphas to obtain (default: 40)
+
+    Returns
+    -------
+    all_alphas : list
+        List containing the different alpha values
+    """
+    # Values taken from the publication
+    last_alpha = init_alpha
+    all_alphas = [last_alpha]
+    # Start loop to get all alpha values
+    for i in range(2, n_alphas + 1):
+        new_alpha = 2 ** (1 / np.log(i) ** 2) * last_alpha
+        all_alphas.append(new_alpha)
+        last_alpha = new_alpha
+    return all_alphas
+
+
+def _dkm_get_probs(squared_diffs: torch.Tensor, alpha: float) -> torch.Tensor:
+    """
+    Predict soft cluster labels given embedded samples.
+
+    Parameters
+    ----------
+    squared_diffs : torch.Tensor
+        the squared distances between points and centers
+    alpha : float
+        the alpha value
+
+    Returns
+    -------
+    prob : torch.Tensor
+        The predicted soft labels
+    """
+    # Shift distances for exponent with min value to avoid underflow (see original implementaion: https://github.com/MaziarMF/deep-k-means/blob/master/compgraph.py)
+    shifted_squared_diffs = squared_diffs - squared_diffs.min(1)[0].reshape((-1, 1))
+    exponent = torch.exp(-alpha * (shifted_squared_diffs))
+    param_softmax = exponent / exponent.sum(1).reshape((-1, 1))
+    return param_softmax
+
+
+class _DKM_Module(torch.nn.Module):
+    """
+    The _DKM_Module. Contains most of the algorithm specific procedures like the loss and prediction functions.
+
+    Parameters
+    ----------
+    init_centers : np.ndarray
+        The initial cluster centers
+    alphas : list
+        list of different alpha values used for the prediction
+    augmentation_invariance : bool
+        If True, augmented samples provided in will be used to learn 
```

Comment:
"in ... will", there is a word missing (also in dipdeck.py). For ENRC I wrote:

 If True, augmented samples provided in custom_dataloaders[0] will be used to learn
        cluster assignments that are invariant to the augmentation transformations

Reply:
Will be fixed in dkm.py and dec.py
I can't see where in dipdeck.py a word is missing.